src/output-json-anomaly.c

```diff
@@ -134,6 +135,137 @@ static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *
     return TM_ECODE_OK;
 }
 
+static int AnomalyAppLayerDecoderEventJson(JsonAnomalyLogThread *aft,
+                        const Packet *p, AppLayerDecoderEvents *decoder_events,
+                        bool is_applayer, const char *layer, uint64_t tx_id)
+{
+    const char *alprotoname = AppLayerGetProtoName(p->flow->alproto);
+
+    SCLogDebug("decoder_events %p event_count %d (last logged %d) %s",
+                decoder_events, decoder_events->cnt,
+                decoder_events->event_last_logged,
+                tx_id != TX_ID_UNUSED ? "tx" : "no-tx");
+
+    for (int i = decoder_events->event_last_logged; i < decoder_events->cnt; i++) {
+        MemBufferReset(aft->json_buffer);
+
+        json_t *js;
+        if (tx_id != TX_ID_UNUSED) {
+            js = CreateJSONHeaderWithTxId(p, LOG_DIR_PACKET,
+                                          ANOMALY_EVENT_TYPE, tx_id);
+        } else {
+            js = CreateJSONHeader(p, LOG_DIR_PACKET, ANOMALY_EVENT_TYPE);
+        }
+        if (unlikely(js == NULL)) {
+            return TM_ECODE_OK;
+        }
+
+        json_t *ajs = json_object();
+        if (unlikely(ajs == NULL)) {
+            json_decref(js);
+            return TM_ECODE_OK;
+        }
+
+        JsonAddCommonOptions(&aft->json_output_ctx->cfg, p, p->flow, js);
+
+        /* Use app layer proto name if available */
+        json_object_set_new(ajs, "alproto",
```

Comment:
this is ```app_proto``` in eve. See JsonAddFlow for example.

Also, you can skip this numeric fallback. The compiler will yell at us if we forgot to keep AppLayerGetProtoName current. Worst case a null should be handled correctly by ```json_string()```

Reply:
> ```
>     "event_no": "2 (of 5)",
> ```
> 
> This looks a little suspicious. Would this value ever be parsed, and is it useful in event visualization? If its purely an informative string, then probably fine. But if its useful to be machine parsed, it might be better to have "event_no" and some other field designated to the total events.
> 

The intent was to be an informational string. Do you think that having this information would be helpful for the consumers who are machine parsing these records? If so, then perhaps an alternate presentation would be better.

> I also see:
> 
> ```
>     "layer": "app_layer"
> ```
> 
> and
> 
> ```
>     "layer": "applayer_parser"
> ```
> 
> Perhaps the first one should just be `applayer`.

Will update.

Reply:
> this is `app_proto` in eve. See JsonAddFlow for example.
> 
> Also, you can skip this numeric fallback. The compiler will yell at us if we forgot to keep AppLayerGetProtoName current. Worst case a null should be handled correctly by `json_string()`

Will use `app_proto` and remove the numeric fallback.